cmd/mps-control-daemon/mps/root.go

```diff
@@ -0,0 +1,53 @@
+/**
+# Copyright 2024 NVIDIA CORPORATION
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package mps
+
+import (
+	"path/filepath"
+
+	spec "github.com/NVIDIA/k8s-device-plugin/api/config/v1"
+)
+
+const (
+	ContainerRoot = Root("/mps")
+)
+
+// Root represents an MPS root.
+// This is where per-resource pipe and log directories are created.
+// For containerised applications the host root is typically mounted to /mps in the container.
+type Root string
+
+// PipeDir returns the per-resource pipe dir for the specified root.
+func (r Root) PipeDir(resourceName spec.ResourceName) string {
+	return r.Path(string(resourceName), "pipe")
+}
+
+// LogDir returns the per-resource pipe dir for the specified root.
+func (r Root) LogDir(resourceName spec.ResourceName) string {
+	return r.Path(string(resourceName), "log")
+}
+
```

Comment:
The issue is that we can't have a per-resource shm. The shm MUST exist at `/dev/shm` in the container and as such it is not rooted at the mps Root. (I suppose technically, the mount is created at `/mps/shm` in the init container (`/run/nvidia/mps/shm` on the host) and mounted to `/dev/shm` in both the daemonset container and the device plugin container.

Maybe what I'm trying to say, is that although symmetry would be nice, it isn't possible due to the fundamental differences between shm and the other two dirs.

Reply:
I understand that, but why can't `plugin.mpsDaemon.ShmDir()` return `/dev/shm` and `plugin.mpsRoot.ShmDir(resource)` return r.Path("shm")?

I'm not talking about symmetry on the location that gets returned, but rather symmetry on the functions that are called to get the info that we need.

Reply:
OK. That's fair. I can update.

Reply:
I have updated. I'm not 100% happy with requiring a resource name to `Root.ShmDir()` since it may give the caller the impression that these are different accross resources. Not a blocker though.

Reply:
It doesn't give that impression to me, necessarily. It just says -- for this resource, what is the `shm` directory. The fact that it happens to be the same (or happens to be different) for all resources shouldn't matter to the caller.